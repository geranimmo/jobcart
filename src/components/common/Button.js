import React from "react";

const Button = props => {
	return <button type="button" {...props}>{props.value}</button>;
};

export { Button };
