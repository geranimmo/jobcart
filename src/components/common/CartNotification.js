import React from "react";

const CartNotification = props => {
	const {
		totalItems,
		className
	} = props;
	const cartTotal = totalItems > 9 ? '9+' : totalItems;
    
	return <div className={`${className} ${totalItems === 0 ? 'hidden' : ''}`}>{cartTotal}</div>;
};

export { CartNotification };
