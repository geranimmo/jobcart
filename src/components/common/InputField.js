import React from "react";

const InputField = props => {
	return <input {...props}/>;
};

export { InputField };
