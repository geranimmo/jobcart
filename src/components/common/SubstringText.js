const SubstringText = props => {
	return props.text.length > props.textMaxLength ?
		props.text.substring(0, props.textMaxLength)+'...'
		:
		props.text;
};

export { SubstringText };