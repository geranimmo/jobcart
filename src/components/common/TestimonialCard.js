import React from 'react';

const TestimonialCard = props => {
	const { TestimonialWrapper } = styles;
    
	return (
		<div style={TestimonialWrapper} className={props.className}>
			{props.children}
		</div>
	);
};

// sample of inline style
const styles = {
	TestimonialWrapper: {
		display: 'inline-block'
	}
};

export { TestimonialCard };