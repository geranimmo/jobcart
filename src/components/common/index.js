export { InputField } from './InputField';
export { Button } from './Button';
export { CartFooter } from './CartFooter';
export { DateTimeConverter } from './DateTimeConverter';
export { SubstringText } from './SubstringText';
export { CartItem } from './CartItem';
export { CartNotification } from './CartNotification';