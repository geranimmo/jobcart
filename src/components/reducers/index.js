import { combineReducers } from 'redux';
import LoginReducer from './LoginReducers';
import ProfileReducer from './ProfileReducers';
import ShoppingCartReducer from './ShoppingCartReducers';
import PackageReducer from './PackageReducers';
import TotalCostReducer from './TotalCostReducers';
import TotalDiscountReducer from './TotalDiscountReducers';

export default combineReducers({
	UserLogin: LoginReducer,
	Profile: ProfileReducer,
	ShoppingCart: ShoppingCartReducer,
	Packages: PackageReducer,
	TotalCost: TotalCostReducer,
	TotalDiscount: TotalDiscountReducer
});
