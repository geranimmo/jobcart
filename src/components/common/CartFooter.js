import React from 'react';
import { Button } from './Button';

const CartFooter = props => {
	let fixedCost = parseFloat((props.totalCartCost - props.totalCartDiscount).toFixed(2));
	
	const buttonProps = {
		className: 'button__checkout',
		value: `Checkout (${props.myCartList.length})`
	};
	
	return (
		<div className={props.className}>
			<div>
				<span>
					<label>Total Cost:</label>
					${fixedCost}
				</span>
				<Button {...buttonProps}/>
			</div>
		</div>
	);
};

export { CartFooter };