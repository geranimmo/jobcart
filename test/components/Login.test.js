import React from 'react';
import { Provider } from 'react-redux';
import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import expect from 'expect';
import { mount, shallow } from 'enzyme';
import { BrowserRouter as Router } from 'react-router-dom';
import renderer from 'react-test-renderer';
import ConnectedLogin, { Login, validateEmail } from '../../src/components/login';
import { InputField } from '../../src/components/common/InputField';
import { Button } from '../../src/components/common/Button';
import reducers from '../../src/components/reducers';

describe('>>> L O G I N ---- Test & Snapshot <<<', () => {
	const store = createStore(reducers, {}, applyMiddleware(thunk));
	const initialProps = { formId: 'signInForm' };
	const usernameVal = 'client@default.com';
	const passwordVal = 'default123';

	let wrapper, shallWrapper;

	beforeEach(() => {
		wrapper = mount(
			<Provider store={store}>
				<Router>
					<ConnectedLogin {...initialProps}/>
				</Router>
			</Provider>
		);
		shallWrapper = shallow(
			<Login {...initialProps}/>
		);
	});

	it('+++ Render the Login component should display all child components', () => {
		const finder = wrapper.find('form.sign__in__form');
		expect(finder.length).toEqual(1);

		const findFormWrapper = finder.find('.form__wrapper');
		expect(findFormWrapper.length).toEqual(4);

		const findInputUsername = findFormWrapper.find('input[name="username"]');
		expect(findInputUsername.length).toEqual(1);

		const findInputPassword = findFormWrapper.find('input[name="password"]');
		expect(findInputPassword.length).toEqual(1);

		const findSubmitButton = findFormWrapper.find('button');
		expect(findSubmitButton.length).toEqual(1);
	});

	it('+++ Contains props that assigned to the Login component', () => {
		expect(wrapper.find('form').prop('id')).toEqual(initialProps.formId);
	});

	it('+++ Capturing Snapshot of Login component +++', () => {
		const renderedValue = renderer.create(
			<Provider store={store}>
				<Router>
					<ConnectedLogin {...initialProps}/>
				</Router>
			</Provider>
		).toJSON();
		
		expect(renderedValue).toMatchSnapshot();
	});
       
	it('+++ Should validate email address format if valid will return true +++', () => {
		expect(validateEmail('test@email.com')).toBe(true); // should return true
		expect(validateEmail('test_email_com')).toBe(false); // should return false
	});

	it('+++ Simulate form password and handlePasswordField() should return value in state +++', () => {
		const finder = shallWrapper.find('form.sign__in__form');
		const findFormWrapper = finder.find('.form__wrapper');
		const findInputPassword = findFormWrapper.find(InputField).at(1).dive();

		findInputPassword.simulate('change', { target: { value: passwordVal } });
		wrapper.instance().forceUpdate();
		wrapper.update();

		expect(shallWrapper.state('passwordValue')).toEqual(passwordVal);
	});

	it('+++ Simulate form username and handleUsernameField() should return value in state +++', () => {
		const finder = shallWrapper.find('form.sign__in__form');
		const findFormWrapper = finder.find('.form__wrapper');
		const findInputUsername = findFormWrapper.find(InputField).at(0).dive();

		findInputUsername.simulate('change', { target: { value: usernameVal } });
		wrapper.instance().forceUpdate();
		wrapper.update();

		expect(shallWrapper.state('usernameValue')).toEqual(usernameVal);
	});

	it('+++ Should show alert window if username or password is null +++', () => {
		window.alert = jest.fn();
		const finder = shallWrapper.find('form.sign__in__form');
		const findFormWrapper = finder.find('.form__wrapper');
		const findSubmitButton = findFormWrapper.find(Button).dive();

		findSubmitButton.simulate('click');
		wrapper.instance().forceUpdate();
		wrapper.update();
		
		expect(window.alert).toHaveBeenCalledWith('Invalid username or password');
		expect(shallWrapper.state('usernameValue')).toEqual(null);
		expect(shallWrapper.state('passwordValue')).toEqual(null);
	});

	it('+++ Should handle enter key when login +++', () => {
		window.alert = jest.fn();
		const finder = shallWrapper.find('form.sign__in__form');
		const findFormWrapper = finder.find('.form__wrapper');
		const findInputUsername = findFormWrapper.find(InputField).at(0).dive();

		findInputUsername.simulate("keypress", { key: 'Enter' });
		wrapper.instance().forceUpdate();
		wrapper.update();
		
		expect(window.alert).toHaveBeenCalledWith('Invalid username or password');
		expect(shallWrapper.state('usernameValue')).toEqual(null);
		expect(shallWrapper.state('passwordValue')).toEqual(null);
	});

	it('+++ Should initialize loginFetch() if username or password is valid +++', () => {
		const history = { push: jest.fn() };
		const loginFetch = jest.fn();
		const wrappers = shallow(
			<Login {...initialProps} history={history} loginFetch={loginFetch}/>
		);
		const pathChangeSpy = jest.spyOn(wrappers.instance(), 'createDataSource');
		const finder = wrappers.find('form.sign__in__form');
		const findFormWrapper = finder.find('.form__wrapper');
		const findInputUsername = findFormWrapper.find(InputField).at(0).dive();
		const findInputPassword = findFormWrapper.find(InputField).at(1).dive();
		const findSubmitButton = findFormWrapper.find(Button).dive();

		findInputUsername.simulate('change', { target: { value: usernameVal } });
		findInputPassword.simulate('change', { target: { value: passwordVal } });

		wrappers.instance().forceUpdate();
		wrappers.update();

		expect(wrappers.state()).toEqual(
			expect.objectContaining({
				usernameValue: usernameVal,
				passwordValue: passwordVal
			})
		);

		findSubmitButton.simulate('click');

		wrappers.instance().forceUpdate();
		wrappers.update();
		
		expect(loginFetch).toHaveBeenCalledWith(
			expect.objectContaining({
				username: usernameVal,
				password: passwordVal
			})
		);

		// handle change path if login success
		wrappers.setProps({ UserLogin: true });
		wrappers.instance().forceUpdate();
		wrappers.update();

		expect(pathChangeSpy.mock.calls.length).toBe(1);
	});

	it('+++ Should initialize componentWillMount() +++', () => {
		const history = { push: jest.fn() };
		const wrappers = shallow(
			<Login {...initialProps} UserLogin={true} history={history}/>
		);

		wrappers.instance().forceUpdate();
		wrappers.update();
		expect(history.push).toHaveBeenCalledWith('/home');
	});
});