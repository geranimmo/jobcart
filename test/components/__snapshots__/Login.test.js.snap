// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`>>> L O G I N ---- Test & Snapshot <<< +++ Capturing Snapshot of Login component +++ 1`] = `
<div
  className="float__container"
>
  <header
    className=""
    id="header"
  >
    <nav
      className="nav__left"
    >
      <a
        aria-current="page"
        className="active"
        href="/"
        onClick={[Function]}
        style={Object {}}
      >
        <div
          className="seek__logo"
        />
      </a>
      
    </nav>
    
  </header>
  <main
    className="login__component"
  >
    <div
      className="content__wrapper"
    >
      <form
        className="sign__in__form"
        id="signInForm"
      >
        <h1>
          Sign in
        </h1>
        <div
          className="form__wrapper"
        >
          <label
            htmlFor="username"
          >
            Email Address
          </label>
          <input
            autoComplete="off"
            id="username"
            maxLength={50}
            minLength={6}
            name="username"
            onChange={[Function]}
            onKeyPress={[Function]}
            type="text"
          />
        </div>
        <div
          className="form__wrapper"
        >
          <label
            htmlFor="password"
          >
            Password
          </label>
          <input
            autoComplete="off"
            id="password"
            maxLength={50}
            minLength={6}
            name="password"
            onChange={[Function]}
            onKeyPress={[Function]}
            type="text"
          />
        </div>
        <div
          className="form__wrapper"
        >
          <button
            className="button__sign__in"
            onClick={[Function]}
            type="button"
            value="Sign in"
          >
            Sign in
          </button>
        </div>
        <div
          className="form__wrapper"
        >
          Don't have account? 
          <b>
            Sign up
          </b>
        </div>
      </form>
    </div>
  </main>
</div>
`;
